File: src/commands/download.rs
use std::path::PathBuf;
use std::process::Command;

use clap::Args;

#[derive(Debug, Args, Clone)]
pub struct DownloadArgs {
    #[arg(help = "URL of the video to download")]
    pub url: String,

    #[arg(short = 'o', long = "output", help = "Output file path")]
    pub output: PathBuf,

    #[arg(long = "audio-only", help = "Download audio only (MP3)")]
    pub audio_only: bool,

    #[arg(
        long = "format",
        default_value = "mp4",
        help = "Output format (mp4, webm, mp3, wav, etc.)"
    )]
    pub format: String,

    #[arg(
        long = "quality",
        help = "Video quality (best, worst, 720p, 1080p, etc.)"
    )]
    pub quality: Option<String>,

    #[arg(long, help = "Show download progress")]
    pub verbose: bool,

    #[arg(long = "timeout", default_value = "600", help = "Timeout in seconds")]
    pub timeout: u64,
}

pub fn run(cli: DownloadArgs) -> Result<(), String> {
    if cli.verbose {
        eprintln!("mpipe: downloading from {}", cli.url);
    }

    let output_path = cli.output;
    let output_path_str = output_path
        .to_str()
        .ok_or_else(|| "Invalid output path".to_string())?;

    let mut args = vec![
        "--output".to_string(),
        output_path_str.to_string(),
        "--no-part".to_string(),
        "--no-clean-info-json".to_string(),
    ];

    if cli.audio_only {
        args.push("--extract-audio".to_string());
        args.push("--audio-format".to_string());
        args.push(cli.format.clone());
    } else {
        args.push("--format".to_string());
        if let Some(quality) = &cli.quality {
            args.push(format!("bestvideo[height<={}]+bestaudio/best", quality));
        } else {
            args.push("bestvideo+bestaudio/best".to_string());
        }
        args.push("--merge-output-format".to_string());
        args.push("mp4".to_string());
    }

    args.push(cli.url.clone());

    if cli.verbose {
        eprintln!("mpipe: running yt-dlp with args: {:?}", args);
    }

    let output = Command::new("yt-dlp")
        .args(&args)
        .output()
        .map_err(|e| format!("Failed to run yt-dlp: {}", e))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("yt-dlp failed: {}", stderr));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    if cli.verbose {
        eprintln!("mpipe: {}", stdout);
    }

    if output_path.exists() {
        println!("{}", output_path.display());
        Ok(())
    } else {
        Err(format!("Output file not found: {}", output_path.display()))
    }
}

File: mpipe/commands/download.py
# SPDX-License-Identifier: GPL-3.0-or-later
#
# mpipe - Multi-provider LLM CLI tools
# Copyright (C) 2026 Nicolas Schaetti and contributors
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

"""mpipe.commands.download module.

Notes
-----
This module is part of the Python port of the `mpipe` project.
"""
from __future__ import annotations

import subprocess
from pathlib import Path

import click

from mpipe.console import console, err_console


@click.command("download")
@click.argument("url")
@click.option("-o", "--output", type=click.Path(path_type=Path), required=True)
@click.option("--audio-only", is_flag=True)
@click.option("--format", "fmt", default="mp4")
@click.option("--quality")
@click.option("--verbose", is_flag=True)
@click.option("--timeout", default=600, type=int)
def download_command(
    url: str,
    output: Path,
    audio_only: bool,
    fmt: str,
    quality: str | None,
    verbose: bool,
    timeout: int,
) -> None:
    """Download command.

    Parameters
    ----------
    url : str
        Argument value.
    output : Path
        Argument value.
    audio_only : bool
        Argument value.
    fmt : str
        Argument value.
    quality : str | None
        Argument value.
    verbose : bool
        Argument value.
    timeout : int
        Argument value.

    Returns
    -------
    None
        Returned value.
    """
    del timeout
    if verbose:
        err_console.print(f"mpipe: downloading from {url}")
    # end if

    args = [
        "--output",
        str(output),
        "--no-part",
        "--no-clean-info-json",
    ]
    if audio_only:
        args.extend(["--extract-audio", "--audio-format", fmt])
    else:
        args.append("--format")
        args.append(f"bestvideo[height<={quality}]+bestaudio/best" if quality else "bestvideo+bestaudio/best")
        args.extend(["--merge-output-format", "mp4"])
    # end if
    args.append(url)

    if verbose:
        err_console.print(f"mpipe: running yt-dlp with args: {args}")
    # end if

    process = subprocess.run(["yt-dlp", *args], capture_output=True, text=True)
    if process.returncode != 0:
        raise click.ClickException(f"yt-dlp failed: {process.stderr}")
    # end if

    if verbose and process.stdout:
        err_console.print(f"mpipe: {process.stdout}")
    # end if
    if output.exists():
        console.print(str(output))
        return
    # end if
    raise click.ClickException(f"Output file not found: {output}")
# end def download_command
